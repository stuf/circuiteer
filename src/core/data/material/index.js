/**
 * Material lookups
 */

/**
 *
 */
const natural = {
  soil: {},
  organic: {},
  compound: {},
  resin: {},
};

const refined = {
  coal: {
    from: 'organic',
  },
};

const composite = {
  rubber: {
    from: ['organic', 'resin'],
  },
  plastic: {
    from: ['compound', 'coal'],
  },
};

const atmospheric = {
  hydrogen: {},
};
