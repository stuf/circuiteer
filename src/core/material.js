/**
 * @module
 * @deprecated
 */

/**
 * Material type enum
 */
export const MaterialType = {
  NATURAL: 'natural',
  REFINED: 'refined',
  COMPOSITE: 'composite',
  ATMOSPHERIC: 'atmospheric',
  OTHER: 'other',
};

/**
 * @deprecated
 */
const materials = {
  [MaterialType.NATURAL]: [
    {
      id: 'soil',
      displayName: 'Soil',
    },
    {
      id: 'organic',
      displayName: 'Organic',
    },
    {
      id: 'compound',
      displayName: 'Compound',
    },
    {
      id: 'resin',
      displayName: 'Resin',
    },
  ],

  [MaterialType.REFINED]: [
    {
      id: 'copper',
      displayName: 'Copper',
    },
    {
      id: 'glass',
      displayName: 'Glass',
    },
  ],

  [MaterialType.COMPOSITE]: [],

  [MaterialType.ATMOSPHERIC]: [],

  [MaterialType.OTHER]: [],
};

/**
 * @deprecated
 */
export const Material = Object.entries(materials).reduce((o, [k, xs]) => {
  return Object.assign(
    {},
    o,
    xs.reduce((_o, x) => ({ ..._o, [x.id]: { ...x, type: k } }), {}),
  );
}, {});
