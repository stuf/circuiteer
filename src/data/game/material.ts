import { stringRepr } from 'common/meta';
import { ILocation } from './location';

export enum MaterialType {
  NATURAL,
  REFINED,
  COMPOSITE,
  ATMOSPHERIC,
  OTHER,
}

export interface IBaseMaterial {
  type: MaterialType;
  id: string;
  name?: string;
  location?: ILocation[];
}

export class BaseMaterial implements IBaseMaterial {
  typeName = 'BaseMaterial';
  type: MaterialType;
  id: string;
  name?: string;
  location?: ILocation[];

  constructor(obj: IBaseMaterial) {
    this.type = obj.type;
    this.id = obj.id;
    this.name = obj.name;
    this.location = obj.location;
  }

  toObject() {
    return {
      type: this.type,
      id: this.id,
      name: this.name,
      location: this.location,
    };
  }

  toJSON() {
    return this.toObject();
  }

  toString() {
    return stringRepr(this.typeName, this.toObject());
  }
}

export const mkBaseMaterial = (obj: IBaseMaterial) => new BaseMaterial(obj);
