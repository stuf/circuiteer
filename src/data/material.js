import * as I from 'infestines';
import * as L from 'partial.lenses';

export function BaseMaterial(object) {
  if (!I.isInstanceOfU(BaseMaterial, this)) return new BaseMaterial(object);

  this._object = Object.assign({}, object);
}

//

export const NaturalMaterial = I.inherit(function NaturalMaterial(object) {
  if (!I.isInstanceOfU(NaturalMaterial, this))
    return new NaturalMaterial(object);

  BaseMaterial.call(this);
}, BaseMaterial);

//

export const RefinedMaterial = I.inherit(function RefinedMaterial(object) {
  if (!I.isInstanceOfU(RefinedMaterial, this))
    return new RefinedMaterial(object);

  BaseMaterial.call(this);
}, BaseMaterial);

//

export const CompositeMaterial = I.inherit(function CompositeMaterial(object) {
  if (!I.isInstanceOfU(CompositeMaterial, this))
    return new CompositeMaterial(object);

  BaseMaterial.call(this);
}, BaseMaterial);

//

export const AtmosphericMaterial = I.inherit(function AtmosphericMaterial(
  object,
) {
  if (!I.isInstanceOfU(AtmosphericMaterial, this))
    return new AtmosphericMaterial(object);

  BaseMaterial.call(this);
},
BaseMaterial);
