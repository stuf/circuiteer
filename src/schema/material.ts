export enum MaterialType {
  NATURAL = 'natural',
  REFINED = 'refined',
  COMPOSITE = 'composite',
  ATMOSPHERIC = 'atmospheric',
  OTHER = 'other',
}

export class Material {
  id: string;
  name: string;
  type: MaterialType;

  constructor(id, name, type) {
    this.id = id;
    this.name = name;
    this.type = type;
  }

  toJSON() {
    return {
      id: this.id,
      name: this.name,
      type: this.type,
    };
  }
}

export class NaturalMaterial extends Material {
  constructor(id, name) {
    super(id, name, MaterialType.NATURAL);
  }
}

export class RefinedMaterial extends Material {
  constructor(id, name) {
    super(id, name, MaterialType.REFINED);
  }
}

export class CompositeMaterial extends Material {
  constructor(id, name) {
    super(id, name, MaterialType.COMPOSITE);
  }
}

export class AtmosphericMaterial extends Material {
  constructor(id, name) {
    super(id, name, MaterialType.ATMOSPHERIC);
  }
}

export class OtherMaterial extends Material {
  constructor(id, name) {
    super(id, name, MaterialType.OTHER);
  }
}
