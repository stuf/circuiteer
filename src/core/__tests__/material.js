import * as Mat from '../material';

test('Material', () => {
  expect(Mat).toMatchSnapshot();
});
