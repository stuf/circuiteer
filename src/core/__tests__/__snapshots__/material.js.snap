// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Material 1`] = `
Object {
  "Material": Object {
    "compound": Object {
      "displayName": "Compound",
      "id": "compound",
      "type": "natural",
    },
    "copper": Object {
      "displayName": "Copper",
      "id": "copper",
      "type": "refined",
    },
    "glass": Object {
      "displayName": "Glass",
      "id": "glass",
      "type": "refined",
    },
    "organic": Object {
      "displayName": "Organic",
      "id": "organic",
      "type": "natural",
    },
    "resin": Object {
      "displayName": "Resin",
      "id": "resin",
      "type": "natural",
    },
    "soil": Object {
      "displayName": "Soil",
      "id": "soil",
      "type": "natural",
    },
  },
  "MaterialType": Object {
    "ATMOSPHERIC": "atmospheric",
    "COMPOSITE": "composite",
    "NATURAL": "natural",
    "OTHER": "other",
    "REFINED": "refined",
  },
}
`;
